ase/sw/app_backend.c

```diff
@@ -1340,56 +1342,76 @@ bool deallocate_buffer_by_index(int search_index)
 {
     FUNC_CALL_ENTRY;
     bool value;
-    uint64_t *bufptr = (uint64_t *) NULL;
-    struct wsmeta_t *wsptr;
+	struct buffer_t *bufptr = (struct buffer_t *) NULL;
+    struct buffer_t *ptr;
 
     ASE_MSG("Deallocate request index = %d ... \n", search_index);
 
-    // Traverse wsmeta_t
-    wsptr = wsmeta_head;
-    while (wsptr != NULL) {
-	if (wsptr->index == search_index) {
-	    bufptr = wsptr->buf_structaddr;
-	    ASE_DBG("FOUND\n");
-	    break;
-	} else {
-	    wsptr = wsptr->next;
-	}
+    // Traverse buffer list
+    ptr = buf_head;
+    while (ptr != NULL) {
+		if (ptr->index == search_index) {
+			bufptr = ptr;
+			ASE_DBG("FOUND\n");
+			break;
+		} else {
+		  ptr = ptr->next;
+		}
     }
 
-
     // Call deallocate
-    if ((bufptr != NULL) && (wsptr->valid == 1)) {
-	deallocate_buffer((struct buffer_t *) bufptr);
-	wsptr->valid = 0;
-	value = true;
+    if ((bufptr != NULL) && (bufptr->valid == ASE_BUFFER_VALID)) {
+		deallocate_buffer((struct buffer_t *) bufptr);
+		value = true;
     } else {
-	ASE_MSG("Buffer pointer was returned as NULL\n");
-	value = false;
+		ASE_MSG("Buffer pointer was returned as NULL\n");
+		value = false;
     }
 
     FUNC_CALL_EXIT;
     return value;
 }
 
+void free_buffers(void)
+{
+	FUNC_CALL_ENTRY;
+
+	struct buffer_t *bufptr = (struct buffer_t *) NULL;
+	struct buffer_t *ptr;
+
+	ASE_MSG("Deallocate all buffers ... \n");
+
+	// Traverse buffer list
+	ptr = buf_head;
+	while (ptr != NULL) {
+		bufptr = ptr;
+		if (ptr->valid == ASE_BUFFER_VALID)
+			deallocate_buffer(ptr);
+
+		ptr = ptr->next;
+		free(bufptr);
+	}
+
+	FUNC_CALL_EXIT;
+}
+
 
 /*
- * Traverse WSmeta array and find buffer by WSID
+ * Traverse buffer array and find buffer by ID
  */
 struct buffer_t *find_buffer_by_index(uint64_t wsid)
 {
     struct buffer_t *bufptr = (struct buffer_t *) NULL;
-    struct wsmeta_t *trav_ptr;
+    struct buffer_t *trav_ptr;
 
-    trav_ptr = wsmeta_head;
+    trav_ptr = buf_head;
     while (trav_ptr != NULL) {
-	if (trav_ptr->index == wsid) {
-	    bufptr =
-		(struct buffer_t *) trav_ptr->buf_structaddr;
-	    break;
-	} else {
-	    trav_ptr = trav_ptr->next;
-	}
+		if (trav_ptr->index == wsid) {
```

Comment:
Application 'wsid' does not equal trav_ptr->index.
For eg. from simulation log of hello_fpga/NLB0, the DSM, (WSID=0) reads:

**Buffer 2** Allocated  (located /dev/shm//buf0.2671733310964392) =>
                Host App Virtual Addr  = 0x7f6a5d605000
                HW Physical Addr       = 0xc3ca00000
                HW CacheAligned Addr   = 0x30f28000
                Workspace Size (bytes) = 2097152

Reply:
Can you point me to where that differentiation between application WSIDs and ASE workspace indices is made in the original code? I couldn't see it being removed in the diff.

Reply:
I could not see any difference of application WSIDs and buffer index. There is no special handling of index regarding different use cases. At line 1202 in allocate_buffer() function, buffer index was incremented by asebuf_index_count. No matter what the use case of buffers are, MMIO or user buffers, it all invokes allocate_buffer() function to get its index based on global counter of asebuf_index_count.

Reply:
Can you quickly confirm this by doing these steps:
1. Add "#define ASE_DEBUG 1" to top of sw/ase_common.h
2. Rebuild ASE. This will generate a few more log files, one of them will be $ASE_WORKDIR/ase_pagetable.log
3. Correlate the buffer allocation index numbers in ccip_transactions and the file in 2.

Reply:
When I enabled "#define ASE_DEBUG 1", it generated a zero size ase_pagetable.log file. Is there anything missing here?

Reply:
@rahulrs Please provide code evidence for "'wsid' may not necessarily be buffer_t->index". If evidence is not provided by Sunday (11/26) and DCP tests pass, we will merge this pull request branch into master branch.